### PR TITLE
Use BasicLifecycler for Compactor and autoforget unhealthy instances

### LINKED ISSUE #1588: Auto-forget unhealthy compactors
#### Is your feature request related to a problem? Please describe.

When a compactor is not cleanly shutdown, it is not unregistered from the ring. This means that if a compactor is ever uncleanly shutdown when it does not plan on coming back, a subset of tenants (those assigned to the missing instance) will never be compacted. Some example scenarios:

- Compactors are scaled down, but for some reason, the compactor is shutdown uncleanly, leaving e.g. compactor-N in the ring.
- The compactor is renamed, but was shutdown uncleanly before the rename, leaving the old compactor name in the ring.

This will avoid the need for manual intervention to resolve the issue.

#### Describe the solution you'd like

We should use `ring.NewAutoForgetDelegate` on the compactor ring. Compactors are not stateful in any way, so there is no harm in removing them automatically.

#### Describe alternatives you've considered

Do nothing - Compactors can continue to to be manually forgotten

### PR DESCRIPTION
**Pull request by 56quarters**
Signed-off-by: Nick Pillitteri <nick.pillitteri@grafana.com>

#### What this PR does

Change the compactor to use BasicLifecycler instead of Lifecycler so that we can make use of autoforget functionality. This works around an issue where ownership of tenants is retained by unhealthy instances.

The behavior of compactors while starting changes slightly because of the use of BasicLifecycler instead of Lifecycler.

* Lifecycler didn't join the ring when starting, only after running so the `wait_active_instance_timeout` was only needed while waiting for the instance to become active.
* BasicLifecycler joins the ring while starting, thus we need to apply the `wait_active_instance_timeout` when starting the lifecycler and waiting for it to become active in the ring, not _just_ when waiting for the instance to become active.

#### Which issue(s) this PR fixes or relates to

See #3708
Fixes #1588

#### Checklist

- [X] Tests updated
- [ ] Documentation added
- [X] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/compactor/compactor.go [hunk 1]
```diff
@@ -383,42 +392,39 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
...
+	c.ringSubservicesWatcher = services.NewFailureWatcher()
+	c.ringSubservicesWatcher.WatchManager(c.ringSubservices)
+	if err = c.ringSubservices.StartAsync(ctx); err != nil {
+		return errors.Wrap(err, "unable to start compactor ring dependencies")
 	}
 
-	if err != nil {
+	ctxTimeout, cancel = context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
+	defer cancel()
+	if err = c.ringSubservices.AwaitHealthy(ctxTimeout); err != nil {
```
Comment (56quarters) [inline pkg/compactor/compactor.go:413]: I'm not sure if this is the right choice. There was previously no timeout on how long we'd wait for the ring and lifecycler to become healthy. This was fine because the `ring.Lifecycler` doesn't actually have a "start" method and so it isn't doing anything when starting and becoming healthy. However, because we're now using the `ring.BasicLifecycler`, it actually joins the ring when starting and so it [might hit this issue](https://github.com/cortexproject/cortex/issues/3914) if there was no timeout for startup.

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -456,7 +456,7 @@ func newRingAndLifecycler(cfg RingConfig, instanceCount *atomic.Uint32, logger l
 	}
 
 	var delegate ring.BasicLifecyclerDelegate
-	delegate = ring.NewInstanceRegisterDelegate(ring.ACTIVE, ringNumTokens)
```
Comment (pracucci) [inline pkg/distributor/distributor.go:459]: Why these changes in distributor? Can you confirm they're not necessary for this PR?

File: pkg/compactor/compactor.go [hunk 3]
```diff
@@ -458,6 +464,35 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
...
+	delegate = ring.NewInstanceRegisterDelegate(ring.ACTIVE, lifecyclerCfg.NumTokens)
+	delegate = ring.NewLeaveOnStoppingDelegate(delegate, logger)
+	delegate = ring.NewAutoForgetDelegate(ringAutoForgetUnhealthyPeriods*lifecyclerCfg.HeartbeatTimeout, delegate, logger)
+
+	compactorsLifecycler, err := ring.NewBasicLifecycler(lifecyclerCfg, "compactor", ringKey, kvStore, delegate, logger, prometheus.WrapRegistererWithPrefix("cortex_", reg))
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to initialize compactors' lifecycler")
+	}
+
+	compactorsRing, err := ring.NewWithStoreClientAndStrategy(cfg.ToRingConfig(), "compactor", ringKey, kvStore, ring.NewDefaultReplicationStrategy(), prometheus.WrapRegistererWithPrefix("cortex_", reg), logger)
```
Comment (pracucci) [inline pkg/compactor/compactor.go:488]: We're reusing `kvStore` for the ring client too. `kvStore` is created with name `compactor-lifecycler` which doesn't apply if we reuse it for the ring client too. In the old implementation, the name of the KV store used for the client was called `compactor-ring`. I see two options:
1. Use two different KV stores, to keep previous behaviour
2. Rename `kvStore` from `compactor-lifecycler` to just `compactor`, then please double check if there's any dashboard or alert to update

File: pkg/compactor/compactor.go [hunk 4]
```diff
@@ -383,42 +392,39 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
...
 	if err != nil {
-		return errors.Wrap(err, "unable to initialize compactor ring lifecycler")
+		return err
 	}
 
-	c.ring, err = ring.New(lifecyclerCfg.RingConfig, "compactor", CompactorRingKey, c.logger, prometheus.WrapRegistererWithPrefix("cortex_", c.registerer))
+	c.ringSubservices, err = services.NewManager(c.ringLifecycler, c.ring)
 	if err != nil {
-		return errors.Wrap(err, "unable to initialize compactor ring")
+		return errors.Wrap(err, "unable to start compactor ring dependencies")
```
Comment (pracucci) [inline pkg/compactor/compactor.go:402]: [nit] We don't start subservices with the call to `services.NewManager()`, so I would rephrase the error message.

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (pracucci) [reply pkg/compactor/compactor.go:413]: I think applying `WaitActiveInstanceTimeout` here as well is correct.

File: pkg/compactor/compactor.go [hunk 5]
```diff
@@ -383,42 +392,39 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
...
-	// any other component depending on the users scanner, because the
-	// users scanner depends on the ring (to check whether an user belongs
-	// to this shard or not).
+	// If sharding is enabled we should wait until this instance is ACTIVE within the ring. This
+	// MUST be done before starting any other component depending on the users scanner, because
+	// the users scanner depends on the ring (to check whether a user belongs to this shard or not).
 	level.Info(c.logger).Log("msg", "waiting until compactor is ACTIVE in the ring")
 
-	ctxWithTimeout, cancel := context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
+	ctxTimeout, cancel = context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
```
Comment (pracucci) [inline pkg/compactor/compactor.go:422]: I don't think we should create a new `ctxTimeout` here, because we apply `WaitActiveInstanceTimeout` two times, while I think we should just reuse the previous `ctxTimeout`.

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! Overall LGTM. I just left a couple of comments.

Do you mind testing it in a dev cluster and running a rollout to ensure everything works as expected? Unfortunately both unit tests and integration tests are not very good to catch issues with the ring initialization and state changes.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (56quarters) [reply pkg/distributor/distributor.go:459]: They are not needed for this PR, I just noticed that the way the number of tokens constant was being used was inconsistent.

File: pkg/compactor/compactor.go (hunk 3 above)
Comment (56quarters) [reply pkg/compactor/compactor.go:488]: I will take a look at this.

File: pkg/compactor/compactor.go (hunk 5 above)
Comment (56quarters) [reply pkg/compactor/compactor.go:422]: The way `WithTimeout` is implemented, any time spent waiting for the first place we use it takes time away from the second place we use it. If I implement your suggestion, starting and waiting for the instance to become active will only have N seconds total to complete. The way it's currently written, starting has N seconds to complete and waiting for the instance to become active has N seconds to complete. I can make that change but want to confirm that's what you intend.

File: pkg/compactor/compactor.go (hunk 3 above)
Comment (56quarters) [reply pkg/compactor/compactor.go:488]: I'm going to go back to use two different KV stores and maintain the previous behavior rather than have compactors using a different setup than other components.

Comment (56quarters) [reply pkg/compactor/compactor.go:488]: This also exposes a problem with the way the kvstore is created for compactors (via this PR) and distributors: The `prometheus.Registerer` used for the kvstore isn't prefixed with `cortex_` so it emits metrics like `kv_request_duration_seconds_count` compared to ingesters which use `cortex_kv_request_duration_seconds_count`

File: pkg/compactor/compactor.go (hunk 5 above)
Comment (pracucci) [reply pkg/compactor/compactor.go:422]: > I can make that change but want to confirm that's what you intend.

That's what I meant (all in all it mimics the previous behaviour) but I honestly don't feel very strong about it. The whole purpose of this timeout was just to avoid to waiting indefinitely. Keep your code as is 👍

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -445,6 +445,7 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 
 // newRingAndLifecycler creates a new distributor ring and lifecycler with all required lifecycler delegates
 func newRingAndLifecycler(cfg RingConfig, instanceCount *atomic.Uint32, logger log.Logger, reg prometheus.Registerer) (*ring.Ring, *ring.BasicLifecycler, error) {
+	reg = prometheus.WrapRegistererWithPrefix("cortex_", reg)
```
Comment (pracucci) [inline pkg/distributor/distributor.go:448]: This changes the metric names for the KV client used by the distributor. Two comments:
1. Can you double check what's the impact on dashboards and alerts?
2. This should be mentioned in the CHANGELOG entry too

Comment (pracucci) [reply pkg/distributor/distributor.go:448]: > Can you double check what's the impact on dashboards and alerts?

I checked it, and we always query `cortex_kv_request_duration_seconds_*` in dashboards and alerts. This means that this is a fix, not a regression, which is good.

However, querying the metric name `kv_request_duration_seconds` (without the `cortex_` prefix) I've found it's also exposed by the query-scheduler.

**Suggestion**: I suggest to revert this change from this PR (keeping the "broken" version) and open a separate PR to fix it both here in the distributor and query-scheduler, so that the change is better tracked in the commits log.

File: pkg/compactor/compactor_test.go [hunk 7]
```diff
@@ -1758,18 +1762,13 @@ func TestMultitenantCompactor_ShouldFailCompactionOnTimeout(t *testing.T) {
...
 
 	// Try to start the compactor with a bad consul kv-store. The
 	err := services.StartAndAwaitRunning(context.Background(), c)
 
 	// Assert that the compactor timed out
-	assert.Equal(t, context.DeadlineExceeded, err)
-
-	assert.ElementsMatch(t, []string{
-		`level=info component=compactor msg="waiting until compactor is ACTIVE in the ring"`,
-		`level=error component=compactor msg="compactor failed to become ACTIVE in the ring" err="context deadline exceeded"`,
```
Comment (pracucci) [inline pkg/compactor/compactor_test.go:1771]: Why did you delete this log in the `starting()`? I was handy to check the actual reason why it failed to start.

**Review by pracucci — APPROVED** [review]
LGTM! I left a couple of minor comments.

File: pkg/compactor/compactor_test.go (hunk 7 above)
Comment (56quarters) [reply pkg/compactor/compactor_test.go:1771]: Because we no longer emit these log messages in the test now that we've added a timeout to `ringSubservices.AwaitHealthy` and `ring.WaitInstanceState` - these messages only happen after `ringSubservices.AwaitHealthy` has run. I could rearrange things so they're emitted but I'm not a fan of testing log messages.

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (56quarters) [reply pkg/distributor/distributor.go]: Reverted. I'll open a separate PR to fix for distributors and query-scheduler.

File: pkg/compactor/compactor_test.go (hunk 7 above)
Comment (pracucci) [reply pkg/compactor/compactor_test.go:1771]: Ok keep it as is and let's move on 😉
